backend/layers/processing/process_validate_atac.py

```diff
@@ -0,0 +1,219 @@
+            self.logger.warning("Skipping fragment validation")
+            self.update_processing_status(
+                dataset_version_id,
+                DatasetStatusKey.ATAC_FRAGMENT,
+                DatasetConversionStatus.SKIPPED,
+                validation_errors=[str(e)],
+            )
+            return True
+
+        if manifest.atac_fragment is None:
+            if result:  # fragment file required
+                self.update_processing_status(
+                    dataset_version_id, DatasetStatusKey.VALIDATION, DatasetValidationStatus.INVALID
+                )
+                raise ValidationAtacFailed(errors=["Anndata requires fragment file"])
+            else:  # fragment file optional
+                self.logger.info("Fragment is optional and not present. Skipping fragment validation.")
+                self.update_processing_status(
+                    dataset_version_id,
+                    DatasetStatusKey.ATAC_FRAGMENT,
+                    DatasetConversionStatus.SKIPPED,
+                    validation_errors=["Fragment is optional and not present."],
+                )
+                return True
+        return False
+
+    def hash_file(self, file_name: str) -> str:
+        """
+        Hash the file
+        :param file_name: the file to hash
+        :return: the hash
+        """
+        with open(file_name, "rb") as f:
+            return hashlib.sha256(f.read()).hexdigest()
+
+    def process(
+        self,
+        collection_version_id: CollectionVersionId,
+        dataset_version_id: DatasetVersionId,
+        manifest: IngestionManifest,
+        datasets_bucket: str,
+    ):
+        """
+
+        :param collection_version_id:
+        :param dataset_version_id:
+        :param manifest:
+        :param datasets_bucket:
+        :return:
+        """
+        # Download the original dataset files from URI
+        local_anndata_filename = self.download_from_source_uri(
+            source_uri=str(manifest.anndata),
+            local_path=CorporaConstants.ORIGINAL_H5AD_ARTIFACT_FILENAME,
+        )
+
+        if self.skip_atac_validation(local_anndata_filename, manifest, dataset_version_id):
+            return
+
+        # Download the original fragment file from URI
+        local_fragment_filename = self.download_from_source_uri(
+            source_uri=str(manifest.atac_fragment), local_path=CorporaConstants.ORIGINAL_ATAC_FRAGMENT_FILENAME
+        )
+
+        # Hash the original fragment file. This is used to check if the new fragment file is the same as the old
+        # fragment file to avoid uploading the same file multiple times
+        if self.business_logic.is_already_ingested(manifest.atac_fragment):
+            original_fragment_hash = self.hash_file(local_fragment_filename)
+        else:
+            # new fragment file is from a private source, so we must upload it
+            original_fragment_hash = None
```

Comment:
I think it would make sense to store the hash in the database.

Reply:
in what table and field?

Reply:
That's something we can do as a future optimzation